File: 22.백트래킹/15650.N과 M (2)/Main.java
import java.io.*;
import java.util.StringTokenizer;

public class Main {
    static StringBuilder sb = new StringBuilder();
    static int[] arr;
    static int N, M;

    static void backtrack(int depth) {
        if (depth == M) {
            for (int num : arr) {
                sb.append(num).append(" ");
            }
            sb.append("\n");
            return;
        }

        int start = depth == 0 ? 1 : arr[depth - 1] + 1;
        for (int i = start; i <= N; i++) {
            arr[depth] = i;
            backtrack(depth + 1);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        arr = new int[M];

        backtrack(0);
        bw.write(sb.toString());
        bw.close();
        ;
        br.close();
    }
}

File: 22.백트래킹/15650.N과 M (2)/main.py
import sys

input = sys.stdin.readline


def backtrack(n):
    if n == M:
        print(*arr)
        return

    start = 1 if n == 0 else arr[n - 1] + 1
    for i in range(start, N + 1):
        arr[n] = i
        backtrack(n + 1)


if __name__ == "__main__":
    N, M = map(int, input().split())
    arr = [-1] * M

    backtrack(0)
